feature: add plugins support for custom text/template functions

sethmz opened the pull request:
Some users may want to use custom text/template Go functions in their
templates without having to create a CLI app that can be `exec`ed or
fork/recompile tpl. Supporting plugins allows their Go functions to be
invoked directly, as well as maintain a separate lifecycle from tpl.

Commit:
feature: add plugins support for custom text/template functions

Some users may want to use custom text/template Go functions in their
templates without having to create a CLI app that can be `exec`ed or
fork/recompile tpl. Supporting plugins allows their Go functions to be
invoked directly, as well as maintain a separate lifecycle from tpl.

ripta commented on main.go:
```diff
@@ -106,6 +111,30 @@ func main() {
 		}
 	}
+	if soFiles, err := filepath.Glob(path.Join(*plugDir, "*.so")); err != nil {
+		log.Printf("could not search for plugins in directory %q: %v", *plugDir, err)
```
Since `plugDir` is blank by default, if the user specifies a non-empty value, they presumably want plugins. In such a case, this is a real error, so do you mind modifying this (and the other error reporting below) to respect the value of `onError`? I suppose some refactoring might be necessary, lest error-handling takes up 50% of the lines.

sethmz replied:
Good point! I'll refactor.

ripta commented on main.go:
```diff
@@ -106,6 +111,30 @@ func main() {
+				continue
+			}
+			for k, v := range f() {
+				fm[k] = v
```
Interestingly, this implies that not only can functions defined in plugins override predefined functions, but also plugins—presumably loaded in alphanumeric order of plugin filename—that are loaded later may override functions defined in plugins loaded earlier.

Do you have any thoughts about making that an error instead, or perhaps an alternative?

sethmz replied:
Initially, I thought overriding functions would be fine (and possibly useful/desirable), but I can see how users might find this behavior surprising.

Returning an error might be too restrictive, however, as that would require all function names to be unique across all plugins. Since plugins might not come from the same source, that's impossible to gaurentee.

Perhaps plugin functions should be namespaced using `<plugin_name>_` as the prefix, where `<plugin_name>` is the name of the binary?

ripta commented on test/plugins/foo_func_map.go:
```diff
@@ -0,0 +1,16 @@
+
+func foo() string {
+	return fmt.Sprint("foo")
+}
```
I don't think I can comment on a binary file, but can you add the `libfoo_func_map.so` to the gitignore and remove it from the PR?

sethmz replied:
Whoops. I could have sworn I updated the gitignore. Will fix, sorry!

ripta reviewed (changes requested):
@foowars - Thanks for your contribution. I started a review and had some minor comments.

Commit:
plugins: refactor code loading plugins into helper func

- load plugins in a separate function to simplify error handling
- respect `-on-error` flag when loading plugins
- namespace plugin functions based on the plugin's filename, so
  we don't override functions from sprig, tpl, or other plugins
- gitignore plugin binaries

sethmz commented:
@ripta - thanks for the helpful feedback! I've attempted to address your concerns. Let me know what you think. :)

Commit:
plugins: move code out of main.go into separate file

Plugin logic shouldn't clutter up main.go so much. Clean up by moving
it to plugins.go.

ripta committed:
ci: pull-request automation test

ripta committed:
Merge branch 'master' into feature/plugins

ripta committed:
Merge branch 'master' into feature/plugins

ripta commented:
@foowars - Thanks again for the PR, and sorry it took so long to get it merged!
